streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java

```diff
@@ -0,0 +1,153 @@
+import org.apache.kafka.streams.kstream.Merger;
+import org.apache.kafka.streams.kstream.SessionWindows;
+import org.apache.kafka.streams.kstream.Window;
+import org.apache.kafka.streams.kstream.Windows;
+import org.apache.kafka.streams.kstream.internals.graph.ProcessorParameters;
+import java.util.Collections;
+import org.apache.kafka.streams.kstream.internals.graph.StatefulProcessorNode;
+import org.apache.kafka.streams.kstream.internals.graph.StreamsGraphNode;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.state.StoreBuilder;
+
+class CogroupedStreamAggregateBuilder<K, VOut> {
+
+
+    private final InternalStreamsBuilder builder;
+
+    CogroupedStreamAggregateBuilder(final InternalStreamsBuilder builder) {
+        this.builder = builder;
+    }
+
+    <KR, VIn, W extends Window> KTable<KR, VOut> build(
+            final Map<KGroupedStreamImpl<K, ?>, Aggregator<? super K, ? super Object, VOut>> groupPatterns,
+            final Initializer<VOut> initializer,
+            final NamedInternal named,
+            final StoreBuilder<? extends StateStore> storeBuilder,
+            final Serde<KR> keySerde,
+            final Serde<VOut> valSerde,
+            final String queryableName,
+            final Windows<W> windows,
+            final SessionWindows sessionWindows,
+            final Merger<? super K, VOut> sessionMerger) {
+
+        final Collection<? extends AbstractStream<K, ?>> groupedStreams = new ArrayList<>(groupPatterns.keySet());
+        final AbstractStream<K, ?> kGrouped = groupedStreams.iterator().next();
+        groupedStreams.remove(kGrouped);
+        kGrouped.ensureCopartitionWith(groupedStreams);
+
+
+        final Collection<StreamsGraphNode> processors = new ArrayList<>();
+        boolean stateCreated = false;
+        int counter = 0;
+        for (final Entry<KGroupedStreamImpl<K, ?>, Aggregator<? super K, ? super Object, VOut>> kGroupedStream : groupPatterns
+                .entrySet()) {
+            final StatefulProcessorNode statefulProcessorNode = getStatefulProcessorNode(
+                    kGroupedStream.getValue(),
+                    initializer,
+                    named.withName(named.orElseGenerateWithPrefix(builder, CogroupedKStreamImpl.AGGREGATE_NAME) + "-" + counter++),
+                    stateCreated,
+                    storeBuilder,
+                    windows,
+                    sessionWindows,
+                    sessionMerger);
+            stateCreated = true;
+            processors.add(statefulProcessorNode);
+            builder.addGraphNode(kGroupedStream.getKey().streamsGraphNode, statefulProcessorNode);
+        }
+        final String functionName = named.orElseGenerateWithPrefix(builder, CogroupedKStreamImpl.AGGREGATE_NAME);
+        final ProcessorSupplier<K, VOut> passThrough = new PassThrough<>();
+        final StatefulProcessorNode<K, VOut> tableSourceNode =
```

Comment:
This is not a `tableSourceNode` and should be renamed. -- Also, why is this not a stateless processor?

Reply:
related to my earlier misunderstanding about table sources